widgets/text/text.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package text contains a widget that displays textual data.
+package text
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"image"
+	"sync"
+	"unicode"
+
+	"github.com/mum4k/termdash/canvas"
+	"github.com/mum4k/termdash/terminalapi"
+	"github.com/mum4k/termdash/widgetapi"
+)
+
+// Text displays a block of text.
+//
+// Each line of the text is either trimmed or wrapped according to the provided
+// options. The entire text content is either trimmed or rolled up through the
+// canvas according to the provided options.
+//
+// By default the widget supports scrolling of content with either the keyboard
+// or mouse. See the options for the default keys and mouse buttons.
+//
+// Implements widgetapi.Widget. This object is thread-safe.
+type Text struct {
+	// buff contains the text to be displayed in the widget.
+	buff bytes.Buffer
+	// givenWOpts are write options given for the text.
+	givenWOpts givenWOpts
+
+	// scroll tracks scrolling the position.
+	scroll *scrollTracker
+
+	// lastWidth stores the width of the last canvas the widget drew on.
+	// Used to determine if the previous line wrapping was invalidated.
+	lastWidth int
+	// newText indicates if new text was added to the widget.
+	// Used to determine if the previous line wrapping was invalidated.
+	newText bool
+	// lines stores the starting locations in bytes of all the lines in the
+	// buffer. I.e. positions of newline characters and of any calculated line wraps.
+	lines []int
+
+	// mu protects the Text widget.
+	mu sync.Mutex
+
+	// opts are the provided options.
+	opts *options
+}
+
+// New returns a new text widget.
+func New(opts ...Option) *Text {
+	opt := newOptions(opts...)
+	return &Text{
+		givenWOpts: newGivenWOpts(),
+		scroll:     newScrollTracker(opt),
+		opts:       opt,
+	}
+}
+
+// Reset resets the widget back to empty content.
+func (t *Text) Reset() {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.buff.Reset()
+	t.givenWOpts = newGivenWOpts()
+	t.scroll = newScrollTracker(t.opts)
+	t.lastWidth = 0
+	t.newText = true
```

Comment:
Should this be true if there is no content and has been reset (reset kind of gives the impression on new text was added yet).

Reply:
Good point, renamed newText to contentChanged.